File: python/quantcapital/entities/event.py
"""
事件类定义

事件驱动架构的核心，定义系统中流转的各种事件类型。
"""

from datetime import datetime
from dataclasses import dataclass
from typing import Any, Dict, Optional
from enum import Enum


class EventType(Enum):
    """事件类型枚举"""
    MARKET = "market"      # 行情更新事件
    SIGNAL = "signal"      # 策略信号事件  
    ORDER = "order"        # 订单事件
    FILL = "fill"          # 成交事件
    TIMER = "timer"        # 定时事件


@dataclass
class Event:
    """事件基类"""
    type: str              # 事件类型
    timestamp: datetime    # 事件时间
    data: Dict[str, Any]   # 事件数据


@dataclass  
class MarketEvent(Event):
    """行情更新事件"""
    symbol: str            # 标的代码
    bar_data: 'Bar'        # K线数据
    
    def __post_init__(self):
        self.type = EventType.MARKET.value


@dataclass
class SignalEvent(Event):
    """策略信号事件"""
    strategy_id: str       # 策略ID
    symbol: str            # 标的代码
    direction: str         # 方向: 'BUY', 'SELL', 'HOLD'
    strength: float        # 信号强度 0-1
    price: float           # 信号参考价格
    reason: str            # 信号原因
    
    def __post_init__(self):
        self.type = EventType.SIGNAL.value


@dataclass
class OrderEvent(Event):
    """订单事件"""
    order: 'Order'         # 订单对象
    
    def __post_init__(self):
        self.type = EventType.ORDER.value


@dataclass  
class FillEvent(Event):
    """成交事件"""
    fill: 'Fill'           # 成交对象
    
    def __post_init__(self):
        self.type = EventType.FILL.value


@dataclass
class TimerEvent(Event):
    """定时事件"""
    timer_id: str          # 定时器ID
    interval: int          # 间隔(秒)
    
    def __post_init__(self):
        self.type = EventType.TIMER.value

File: src/main/java/com/quantcapital/entities/event/Event.java
package com.quantcapital.entities.event;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.quantcapital.entities.constant.EventType;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

/**
 * 事件基类
 * <p>
 * 所有事件的基础类，包含事件的通用属性。
 * 支持序列化和反序列化，便于事件的持久化和网络传输。
 *
 * @author QuantCapital Team
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "eventType")
@JsonSubTypes({@JsonSubTypes.Type(value = MarketEvent.class, name = "MARKET"),
        @JsonSubTypes.Type(value = SignalEvent.class, name = "SIGNAL"),
        @JsonSubTypes.Type(value = OrderEvent.class, name = "ORDER"),
        @JsonSubTypes.Type(value = FillEvent.class, name = "FILL"),
        @JsonSubTypes.Type(value = TimerEvent.class, name = "TIMER")})
public abstract class Event {

    /**
     * 事件ID，全局唯一
     */
    private String eventId;

    /**
     * 事件类型
     */
    private EventType type;

    /**
     * 事件时间戳
     */
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime timestamp;

    /**
     * 事件关联的标的代码（可选）
     */
    private String symbol;

    /**
     * 事件优先级（数值越小优先级越高）
     */
    private int priority = 0;

    /**
     * 事件扩展数据
     */
    private Map<String, Object> data;

    /**
     * 构造函数
     *
     * @param type      事件类型
     * @param timestamp 事件时间
     */
    protected Event(EventType type, LocalDateTime timestamp) {
        this.eventId = UUID.randomUUID().toString();
        this.type = type;
        this.timestamp = timestamp;
    }

    /**
     * 构造函数（带标的代码）
     *
     * @param type      事件类型
     * @param timestamp 事件时间
     * @param symbol    标的代码
     */
    protected Event(EventType type, LocalDateTime timestamp, String symbol) {
        this(type, timestamp);
        this.symbol = symbol;
    }

    /**
     * 获取事件描述
     *
     * @return 事件描述字符串
     */
    public abstract String getDescription();

    /**
     * 判断事件是否过期
     *
     * @param currentTime    当前时间
     * @param timeoutSeconds 超时秒数
     * @return 是否过期
     */
    public boolean isExpired(LocalDateTime currentTime, long timeoutSeconds) {
        return timestamp.plusSeconds(timeoutSeconds).isBefore(currentTime);
    }

    @Override
    public String toString() {
        return String.format("%s[id=%s, type=%s, timestamp=%s, symbol=%s]", getClass().getSimpleName(), eventId, type,
                timestamp, symbol);
    }
}